### ISSUE TITLE
feat(examples): port mnist to autograd path (sub-task 2 of #5454)

### ISSUE THREAD

**Issue by mvillmow** [mvillmow/ml-odyssey#5558]
Sub-task **2 of 7** of #5454 — port every manual-gradient convnet example onto the autograd path (`Variable` / `GradientTape` / `optimizer.step(params, tape)`).

## Scope

Port `examples/mnist/` (**Simple MLP**) off the manual forward + `*_backward` + `model.update_parameters(lr, ...)` pattern onto the autograd substrate.

- **Approx LOC delta:** -100 / +50
- **Notes:** Should port near-trivially via a Variable matmul chain.

## Prerequisite — satisfied

The blocker #5452 (`feat(autograd): complete autograd substrate for convnet training`) is **CLOSED**. The substrate this port needs is present in `src/projectodyssey/autograd/`:
`variable_conv2d`, `variable_maxpool2d`, `variable_linear`, `variable_flatten`, `variable_cross_entropy` (all in `variable.mojo`), and `tape.backward()` reverse-topological dispatch (`tape.mojo:296`). So this sub-task is **unblocked**.

## Port checklist (per #5454)

1. **Fork** `examples/mnist/` into an autograd variant first — do **not** delete the manual version until tests pass.
2. **Parity smoke run:** confirm loss/accuracy parity with the manual version on a subset (1 epoch), using a genuine in-container run (no fabricated metrics — committed logs are not evidence).
3. **Default optimizer → AdamW** (matches `examples/grok/lenet_emnist/`); keep SGD selectable.
4. Once **all 7 ports + the grok example** are autograd-based (tracked in #5454), a follow-up deletes the manual implementations and the `model.update_parameters(lr, ...)` pattern from the model classes.

## Definition of done

- [ ] Autograd variant of `examples/mnist/` trains via `GradientTape` + `optimizer.step`.
- [ ] Parity smoke run passes in-container (loss/accuracy within tolerance of the manual version).
- [ ] Default optimizer is AdamW; SGD still selectable.
- [ ] Example's tests run in CI (not merely compiled) — mirror the `example-backward-tests` gate pattern where applicable.
- [ ] Manual `*_backward`/`update_parameters` left intact until the tracker-level cleanup step.

## References

- Tracker: #5454 (suggested order 1→…→7; this is **2**).
- Sub-task 1: #5461 (`lenet_emnist` — closed, reference port).
- `src/projectodyssey/autograd/README.md`, `src/projectodyssey/autograd/optimizers.mojo` (`AdamW`).
- #5465 — a real autograd bug (conv2d backward kernel grads vanished in deep stacks) surfaced during the lenet port; watch for it in the deeper convnets.

**Comment by mvillmow** [comment]
# Implementation Plan — port `examples/mnist/` to the autograd path (sub-task 2 of #5454)

> Plan authored against `main` with every cited API **Read on disk** (per repo discipline: assume 100% wrong-rate on unread APIs). Three of the plan's unverified assumptions were resolved before posting — see **§8**.

## 0. Correction to the issue framing (verified against source)

The body calls mnist a **"Simple MLP"** that ports "near-trivially via a Variable matmul chain," and says **#5465 is only a risk for "deeper convnets."** **Both are wrong for this example:**

- `examples/mnist/model.mojo:113` defines `struct SimpleCNN` — **2× conv2d + 2× maxpool2d + flatten + 2× linear** (forward at model.mojo:200–256). Not an MLP; no bare matmul chain.
- The backward pass therefore runs `backward_conv2d` **twice** (dispatched `tape.mojo:283`), so **#5465 is directly in scope**, not sidestepped.

This plan targets the real `SimpleCNN`. mnist's SimpleCNN has **8 params** (conv1 k/b, conv2 k/b, fc1 w/b, fc2 w/b) — a shorter tail than lenet's 10 (no 3rd FC).

## 1. Approach

- **Create** `examples/mnist/train_autograd.mojo` by forking `examples/mnist/train.mojo`, replacing the manual forward+`*_backward`+`model.update_parameters(...)` body with the `Variable`/`GradientTape`/`optimizer.step(params, tape)` pattern from `examples/lenet_emnist/train_autograd.mojo`.
- **Do NOT** modify/delete `examples/mnist/train.mojo` or `model.mojo` (DoD: manual path stays until the tracker-level cleanup step).
- **No `run_train*` wrapper needed** — mnist's smoke matrix row invokes its entry file directly; the CI row can point straight at `train_autograd.mojo`.

## 2. Ground-truth API signatures (all Read from disk)

**`src/projectodyssey/autograd/variable.mojo`:**
```
variable_conv2d(x, weights, bias, mut tape, stride=1, padding=0) -> Variable        # :686
variable_maxpool2d(x, mut tape, kernel_size, stride=0, padding=0) -> Variable       # :735
variable_relu(x, mut tape) -> Variable                                               # :467
variable_flatten(x, mut tape) -> Variable                                            # :598
variable_linear(x, weights, bias, mut tape) -> Variable                              # :645
variable_cross_entropy(logits, targets, mut tape) -> Variable   # :777 (mean-reduced; records only logits)
Variable.__init__(out self, var data: AnyTensor, requires_grad: Bool, mut tape) raises  # :102
Variable.backward(self, mut tape) raises   # :147 (grad=ones_like, calls tape.backward)
```
**`src/projectodyssey/autograd/tape.mojo`:** `GradientTape.__init__(out self)` :151; `.enable()` :157; `.backward(mut self, output_id, output_grad) raises` **:296** (conv2d backward dispatched :283).

**`src/projectodyssey/autograd/optimizers.mojo`:**
```
AdamW.__init__(lr:Float64=0.001, beta1=0.9, beta2=0.999, epsilon=1e-8, weight_decay=0.01)  # :626
SGD.__init__(learning_rate:Float64, momentum:Float64=0.0)                                    # :107
<both>.step(mut parameters: List[Variable], mut tape) raises      # AdamW :663 / SGD :126
<both>.zero_grad(self, mut tape)                                  # AdamW :810 / SGD :205
```
**`src/projectodyssey/utils/arg_parser.mojo` `create_training_parser()` (:446):** already registers `epochs`,`batch-size`,`lr`,`weight-decay`,`max-batches`,`optimizer`(default `"sgd"`), flags `verbose`,`smoke`. So `--smoke`/`--max-batches`/`--optimizer`/`--weight-decay` need no new registration.
**`src/projectodyssey/training/evaluation.mojo`:** `evaluate_model_simple[M: Model](mut model, images, labels, batch_size=100, num_classes=10, verbose=True) -> Float32` (:230) — generic over `Model`; `SimpleCNN` conforms; reuses `model.forward()` (no autograd at eval).

## 3. Exact edits — new file `examples/mnist/train_autograd.mojo`

- **Imports:** swap manual `*_backward` imports for autograd imports (mirror lenet train_autograd.mojo:38–48); keep `SimpleCNN` + arch constants from `model`, and `load_idx_*`/`normalize_images`/`one_hot_encode`/`evaluate_model_simple`/`create_training_parser`.
- **TrainConfig:** add `optimizer: String`; keep `learning_rate` as `Float64` (optimizers want Float64; manual uses Float32 — convert).
- **`train_batch[O: Optimizer](mut model, input, labels, mut optimizer: O, mut tape)`** — **trait-parametrized is CONFIRMED viable (see §8/assumption 1)**. Build 8 trainable Variables (conv1 k/b, conv2 k/b, fc1 w/b, fc2 w/b) + input/labels as `requires_grad=False`; forward chain:
  `conv2d→relu→maxpool2d` ×2 → `flatten` → `linear→relu` → `linear(logits)` → `cross_entropy`; `loss.backward(tape)`; read scalar via `loss.data._data.bitcast[Float32]()[0]` (lenet :212); append all 8 params **in fixed order**, `optimizer.step(params, tape)`, write back in same order, `optimizer.zero_grad(tape)`.
- **`train_epoch`:** fresh `var tape=GradientTape(); tape.enable()` **per batch**; construct optimizer **once** outside the batch loop (AdamW moments persist). **Use `_set_float64`/`_get_float64` batch copy** (as lenet's autograd port does) — NOT the manual `train.mojo:361–367` 1-byte-per-element copy, which silently corrupts float32.
- **Optimizer default = AdamW, SGD selectable:** parser's built-in default is `"sgd"` (arg_parser.mojo:465) but DoD wants AdamW default. Resolve **in code**: treat unset/non-`"sgd"` as AdamW; only `--optimizer sgd` selects SGD. Verify via §4d.
- **`main`:** keep the `--smoke` synthetic-data branch verbatim (train.mojo:437–459 — `[n_smoke,1,28,28]` f32 + raw uint8 labels, skips dataset load), keep real-MNIST loader, per-epoch `evaluate_model_simple`, and the `if not smoke: model.save_weights(...)` guard. Print selected optimizer in the banner.

## 4. Verification — GENUINE in-container runs only

**Anti-fabrication contract (restate in PR):** *committed logs are never evidence; a truthful failure is acceptable, an invented success is not.* Every number below must come from a real `just _run` whose output is pasted into the PR.

- **(a) Compile** `--Werror {{MOJO_TARGET_CPU}}` (the #6413 AVX-512 strip, justfile:52): `just _run "pixi run mojo run --Werror {{MOJO_TARGET_CPU}} -I /workspace/src -I /workspace -I /workspace/examples/mnist -Xlinker -lm examples/mnist/train_autograd.mojo --smoke --max-batches 1 --batch-size 4 --epochs 1"`.
- **(b) Smoke mechanism:** `just _run "just training-smoke-one mnist train_autograd.mojo"` → exit 0 + ≥2 finite loss lines (recipe asserts finiteness, rejects nan/inf). Also the empirical #5465 check (2 stacked convs) — report the real loss sequence.
- **(c) Parity smoke (real, 1 epoch subset):** run manual `train.mojo` and `train_autograd.mojo --optimizer sgd` on the same `--max-batches 20 --batch-size 32 --lr 0.001` real-MNIST subset; paste both loss curves. Inits are random per run (kaiming_uniform, model.mojo:159; ctor threads no seed), so parity = same order of magnitude + decreasing trend, **not** bitwise equality. Also run `--optimizer adamw` (new default) and confirm a decreasing trend.
- **(d) Default check:** run with **no** `--optimizer` flag; banner must print AdamW.

## 5. Wire into CI — recommendation: **YES**

Add to the `training-smoke` matrix in `.github/workflows/comprehensive-tests.yml` (`include:` list, after the existing mnist row):
```
- { example: mnist, entry: train_autograd.mojo }
```
The `training-smoke` job compiles **and runs** each entry (`--smoke --max-batches 3 --batch-size 4 --epochs 1`) and asserts a finite loss — the correct mechanism gate for an autograd entrypoint, and it exercises `backward_conv2d` so **#5465 regressions are caught in CI**. Precedent: lenet_emnist has two matrix rows. Also add `"mnist train_autograd.mojo"` to `training-smoke-all` in `justfile:1412–1421`. Do **not** add to `example-backward-tests` (that gate targets the hand-written backward passes the autograd path replaces).

## 6. Risks

1. **#5465 (conv-backward vanishing grad) — CLOSED/COMPLETED (see §8).** Fixed, but §4b/§4c still validate empirically (defense-in-depth). mnist does NOT sidestep it (2 stacked convs).
2. **Optimizer-default gap** — resolve in code, verify §4d.
3. **"simplecnn" serializer** (`get_model_parameter_names("simplecnn")`, model.mojo:305) — out of scope, skipped in smoke; if the non-smoke parity run errors here, reproduce with manual `train.mojo` and report as pre-existing, don't fix.
4. **Buggy manual byte-copy** (train.mojo:361–367) — must use `_set_float64/_get_float64`, not copy verbatim.
5. **Float32↔Float64 lr** — keep config Float64, convert.

## 7. DoD → verifiable steps

- [ ] Autograd variant trains via GradientTape + optimizer.step → new file (§3) proven by §4b.
- [ ] Parity smoke passes in-container → §4c (real curves, order-of-magnitude + trend).
- [ ] Default AdamW, SGD selectable → §3 logic + §4d + §4c `--optimizer sgd`.
- [ ] Runs in CI (not just compiled) → §5 matrix + `training-smoke-all` rows.
- [ ] Manual `*_backward`/`update_parameters` intact → `git diff` touches only the new file + 2 CI lists.

## 8. Unverified API Assumptions — 3 RESOLVED before posting, rest deferred to implementation

**Resolved now (Read on disk):**
1. ✅ **`Optimizer` trait declares `step(mut self, mut List[Variable], mut GradientTape) raises` and `zero_grad(self, mut GradientTape)`** — confirmed `optimizer_base.mojo:24` (trait body quotes both). → the **trait-parametrized `train_batch[O: Optimizer]` pattern IS viable**; no fallback needed.
2. ✅ **#5465 is CLOSED/COMPLETED** (`gh issue view 5465`) — conv-backward vanishing-grad fixed. Still validated empirically in §4b.
3. ✅ AdamW/SGD `.step`/`.zero_grad` signatures identical → training loop is optimizer-agnostic except construction (optimizers.mojo:663/810, :126/205).

**Deferred to implementation (verify before relying):**

| # | Assumption | Verify at |
|---|---|---|
| 4 | `get_string/get_int/get_float/get_bool` default-fallback semantics on the parsed-args object (does unset `--optimizer` return "sgd"?) | parsed-args type in `utils/arg_parser.mojo` |
| 5 | `AnyTensor._set_float64`/`_get_float64`/`_data.bitcast[Float32]()` signatures used by lenet port | `tensor/any_tensor.mojo` |
| 6 | `one_hot_encode`/`normalize_images`/`load_idx_*` signatures | `data/formats/*.mojo` |
| 7 | `SimpleCNN` satisfies `Model` trait bound of `evaluate_model_simple` | `core/traits.mojo` |
| 8 | `get_model_parameter_names("simplecnn")` returns 8 names in save_weights append order | `training/model_utils.mojo` |

## Critical files
- `examples/mnist/train.mojo` (fork source), `examples/mnist/model.mojo` (SimpleCNN — unchanged)
- `examples/lenet_emnist/train_autograd.mojo` (structural template)
- `src/projectodyssey/autograd/{variable,tape,optimizers,optimizer_base}.mojo`
- `.github/workflows/comprehensive-tests.yml` (training-smoke matrix), `justfile` (training-smoke recipes)
